### PR TITLE
Activity Testing

### PR DESCRIPTION
**Pull request by philipwhiuk**
Right now we have little testing of "Android code" (Activities, Fragments, etc). It's hard to justify requesting it in code review for new functionality because it would be far beyond the scope of the issue being fixed.

But we need to add it to prevent regressions and verify fixes people write.

I'm going to start by writing skeleton tests for MessageCompose, Accounts and MessageList. Once that's in place, we can start request testing for new features and fixes and gradually cover off more functionality.

### CODE REVIEW

**Comment by philipwhiuk** [comment]
I can't reproduce the UnsupportedOperationException :(

File: k9mail/build.gradle [hunk 1]
```diff
@@ -105,6 +105,12 @@ android {
     }
 }
 
+tasks.withType(Test) {
```
Comment (philipwhiuk) [inline k9mail/build.gradle]: This makes it easier debug failing Travis builds.

File: k9mail/src/main/res/layout/actionbar_custom.xml [hunk 2]
```diff
@@ -40,7 +40,7 @@
 
         <ProgressBar
             android:id="@+id/actionbar_progress"
-            style="?android:attr/indeterminateProgressStyle"
+            android:indeterminate="true"
```
Comment (philipwhiuk) [inline k9mail/src/main/res/layout/actionbar_custom.xml]: This modernises the code slightly and works around Robolectric limitations

File: k9mail/src/test/java/com/fsck/k9/controller/MessagingControllerTest.java [hunk 3]
```diff
@@ -150,6 +150,7 @@ public void setUp() throws MessagingException {
 
     @After
     public void tearDown() throws Exception {
+        removeAccountsInPreferences();
```
Comment (philipwhiuk) [inline k9mail/src/test/java/com/fsck/k9/controller/MessagingControllerTest.java]: This ensures we clean up.

File: k9mail/src/main/java/com/fsck/k9/Preferences.java [hunk 4]
```diff
@@ -32,7 +32,7 @@ public static synchronized Preferences getPreferences(Context context) {
 
 
     private Storage mStorage;
-    private Map<String, Account> accounts = null;
+    private HashMap<String, Account> accounts = null;
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/Preferences.java]: This stops us from providing a HashMap that doesn't support `put()` (See `MessagingControllerTest`)

**Comment by cketti** [comment]
Robolectric is nice for having working implementations of simple Android classes, e.g. `Intent`, `Uri`, `TextUtils`. But we absolutely do not want to use it to test complex system interactions or UI code. If you want to write such tests, make them run on Android.
Ideally, we'll move towards using something like MVP to separate behavior from UI code. Then we can test most of the important stuff on the JVM and only need a couple of UI tests that run on an Android device/emulator.

**Comment by philipwhiuk** [comment]
I strongly disagree. UI tests like Espresso are useful and they certainly have their place. But they take freaking ages to run. A few end-to-end espresso tests that cover user scenarios is a good idea. Using Espresso for component testing activities is not IMO.

If you don't want them I'll just keep them on my fork.

**Comment by cketti** [comment]
I don't disagree about Espresso tests. My main point is that Robolectric is not Android. Using it to simulate something complex like the Activity lifecycle is bound to fail eventually. Also, Robolectric is still quite slow when reading the manifest and instantiating the Application class, content providers etc.
Like I said, I believe using a pattern that decouples behavior from Android UI components is the way to go. That way activities/views will become dead simple and hardly need testing. And the behavior part can be tested on the JVM, and, if we put some work into it, even without Robolectric having to provide implementations for simple framework classes. That'll get us very fast and stable tests.

**Pull request closed without merging**